pkg/gardenlet/controller/controllerinstallation/controllerinstallation_control.go

```diff
@@ -353,13 +370,20 @@ func updateConditions(ctx context.Context, gardenClient kubernetes.Interface, co
 	)
 }
 
-func (c *defaultControllerInstallationControl) isResponsible(controllerInstallation *gardencorev1beta1.ControllerInstallation) (bool, error) {
+func (c *defaultControllerInstallationControl) isResponsible(ctx context.Context, cl client.Client, controllerInstallation *gardencorev1beta1.ControllerInstallation) (bool, error) {
 	controllerRegistration, err := c.controllerRegistrationLister.Get(controllerInstallation.Spec.RegistrationRef.Name)
 	if err != nil {
 		return false, err
 	}
 
 	if deployment := controllerRegistration.Spec.Deployment; deployment != nil {
+		if deploymentName := controllerInstallation.Spec.DeploymentRef; deploymentName != nil {
+			controllerDeployment := &gardencorev1beta1.ControllerDeployment{}
+			if err := cl.Get(ctx, kutil.Key(deploymentName.Name), controllerDeployment); err != nil {
+				return false, err
+			}
+			return controllerDeployment.Spec.Type == installationTypeHelm, nil
+		}
```

Comment:
Why is this nested here?
Generally, it was a little confusing to read this because you are checking the `ControllerInstallation`'s `.spec.deploymentRef` here while you are checking the `ControllerRegistration`'s `.spec.deployment.type` below.

As you already read the `ControllerRegistration` with the client, what about only considering its spec and iterating over `.spec.deployment.deploymentRefs` and read the respective `ControllerDeployment`s?

Reply:
> As you already read the ControllerRegistration with the client, what about only considering its spec and iterating over .spec.deployment.deploymentRefs and read the respective ControllerDeployments?

This meant that we'd considered every `ControllerDeployment` in the list. We may neglect this today since only one is allowed but in the future, if more refs will be allowed, we are only interested in the ref that comes from the `ControllerInstallation`, right? Because then we know for sure, it's the deployment that affects the seed the Gardenlet is responsible for.

Reply:
> we are only interested in the ref that comes from the ControllerInstallation, right?

Right, of course. Makes sense, then let's keep it as is. However, the nesting is not needed, or?